KAFKA-5534: `offsetForTimes` result should include partitions with no offset

vahidhashemian opened the pull request:
For topics that support timestamp search, if no offset is found for a partition, the partition should still be included in the result with a `null` offset value. This `KafkaConsumer` method currently excludes such partitions from the result.

vahidhashemian committed:
KAFKA-5534: `offsetForTimes` result should include partitions with no offsets

For topics that support timestamp search, if no offset is found for a partition, the partition should still be included in the result with a `null` offset value.

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/5806/
Test PASSed (JDK 8 and Scala 2.12).

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/5820/
Test PASSed (JDK 7 and Scala 2.11).

ijuma commented on clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java:
```diff
@@ -1489,14 +1489,18 @@ public void resume(Collection<TopicPartition> partitions) {
                             entry.getValue() + ". The target time cannot be negative.");
+                offsetsForTimes.put(partition, null);
             }
-            return fetcher.getOffsetsByTimes(timestampsToSearch, requestTimeoutMs);
```
Why don't we change `fetcher.getOffsetsByTimes()` instead?

vahidhashemian replied:
I made the change here since @hachikuji emphasized on this method in the corresponding JIRA. I'll submit a patch to move the logic to `fetcher.getOffsetsByTimes()`.

ijuma reviewed (commented):
Thanks for the PR. Can you please add a test that verifies that the bug is fixed?

vahidhashemian reviewed (commented):
@ijuma Thanks for reviewing. I'll submit a patch to address your feedback.

vahidhashemian committed:
Moved the logic to `Fetcher.getOffsetsByTimes()` and updated unit tests to cover it

Also, did some cleanup of variable names in `FetcherTest.java`.

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/5937/
Test PASSed (JDK 7 and Scala 2.11).

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/5922/
Test PASSed (JDK 8 and Scala 2.12).

ijuma commented:
@hachikuji, is this what you had in mind?

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java:
```diff
@@ -429,7 +429,11 @@ private void resetOffsets(final Set<TopicPartition> partitions) {
+
         for (Map.Entry<TopicPartition, OffsetData> entry : offsetData.entrySet()) {
             OffsetData data = entry.getValue();
             if (data == null)
```
With this approach, it seems we could drop this case? There is also the insertion of the null for the UNSUPPORTED_FOR_MESSAGE_FORMAT case in `handleListOffsetResponse` that we probably can drop.

vahidhashemian replied:
Thanks for picking those up. I removed the redundant code in the new patch.

vahidhashemian committed:
Removed redundant code

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/6154/
Test PASSed (JDK 7 and Scala 2.11).

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/6138/
Test PASSed (JDK 8 and Scala 2.12).

ijuma commented:
We should add a note to the upgrade notes and I think we can only merge this in trunk as it does change the behaviour.

vahidhashemian commented:
Thanks @ijuma. I'll update to upgrade notes with this change. I assume it's [this file](https://github.com/apache/kafka/blob/trunk/docs/upgrade.html) that needs to be updated.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java:
```diff
@@ -683,8 +685,8 @@ public void onSuccess(ClientResponse response, RequestFuture<Map<TopicPartition,
      *               generally fail the entire future result. The one exception is UNSUPPORTED_FOR_MESSAGE_FORMAT,
-     *               which indicates that the broker does not support the v1 message format and results in a null
-     *               being inserted into the resulting map.
+     *               which indicates that the broker does not support the v1 message format and causes a null value
```
This comment seems no longer accurate. By the way, I've been wondering why we don't use `Map<TopicPartition, OffsetAndTimestamp>` for the request future here. Doesn't seem like `OffsetData` gives us anything extra.

vahidhashemian replied:
Regarding the comment, do you think it's better to remove the part about null value altogether, or have something like "a null offset value is returned for the partition in this case"?

Also, do you suggest the parameter type change to be done as part of this PR? Thanks.

hachikuji replied:
> Regarding the comment, do you think it's better to remove the part about null value altogether, or have something like "a null offset value is returned for the partition in this case"?

It seems like the actual behavior is to only include offsets that have legitimate values. So maybe we can say that?

>Also, do you suggest the parameter type change to be done as part of this PR? Thanks.

Yes, I was thinking that it would simplify the logic above (we can replace the loop with a `putAll` perhaps).

vahidhashemian replied:
I'm trying to understand how this change would work. One difference between `OffsetData` and `OffsetAndTimestamp` is that `OffsetData` supports null timestamps (its `timestamp` type is `Long`, whereas in `OffsetAndTimestamp` the type is `long`). And in the method we may need to set the timestamp to `null` (i.e. for v0 response).

Also, you suggested replacing the loop with a `putAll`. How would partition-level error handling (and raising the appropriate exception) work if we get rid of the loop?

Thanks for clarifying.

hachikuji replied:
Hmm.. that's a fair point. Maybe it won't work. It's a little confusing because we don't have a null check for the timestamp. I guess we assume that the timestamp will not be null because we have are ensured to be dealing with a v1 (or later) ListOffset request. It might be worth adding a comment somewhere so that expectation is not hidden.

vahidhashemian replied:
Based on what I see, v1 or later APIs would not return a null offset, instead they return -1 when offset is unknown. I'll add that to the method comment.

hachikuji replied:
Is that right? The comment is describing the `UNSUPPORTED_FOR_MESSAGE_FORMAT` case. It seems that we now we just leave the partition out of the map now?

vahidhashemian replied:
You're right, what I meant was the `ListOffset` v1 and later doesn't seem to return a null "timestamp" value (ever); instead `-1` is returned when the value is unknown. I didn't mean to associate that with the partition error in the comment, but rather a generic statement that you suggested should be documented somewhere. I hope it makes sense now :)

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java:
```diff
@@ -429,12 +429,14 @@ private void resetOffsets(final Set<TopicPartition> partitions) {
-            if (data == null)
-                offsetsByTimes.put(entry.getKey(), null);
-            else
+            if (data != null)
```
Do we need this check anymore? Are there any cases remaining where we insert a null value?

vahidhashemian replied:
You're right, it's not needed anymore because the partitions with no found offsets will not be in `offsetData`. Thanks.

vahidhashemian committed:
Minor update

- Removed a redundant `if` statement
- Added an item to upgrade docs
- Improved a Javadoc

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/6180/
Test PASSed (JDK 7 and Scala 2.11).

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/6164/
Test PASSed (JDK 8 and Scala 2.12).

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java:
```diff
@@ -429,14 +429,14 @@ private void resetOffsets(final Set<TopicPartition> partitions) {
+            offsetsByTimes.put(entry.getKey(), null);
+
+        for (Map.Entry<TopicPartition, OffsetData> entry : offsetData.entrySet())
+            offsetsByTimes.put(entry.getKey(), new OffsetAndTimestamp(entry.getValue().offset, entry.getValue().timestamp));
```
Can you add a comment here that `entry.getValue().timestamp` will not be null since we are guaranteed to be dealing with a v1 request?

vahidhashemian replied:
Sure, I added it in the new commit.

vahidhashemian committed:
Added a comment to clarify a non-null timestamp value

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/6210/
Test PASSed (JDK 8 and Scala 2.12).

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/6226/
Test PASSed (JDK 7 and Scala 2.11).

hachikuji reviewed (approved):
Thanks for the patch. LGTM
